### PR TITLE
Integrate `kw_bd` to the project and add migration script

### PR DESCRIPTION
**Pull request by davidbtadokoro**
This PR is a continuation of the work made by @kwy95 and @magalilemes in [**src: db: introduce db model and migration script**](https://github.com/kworkflow/kworkflow/pull/531) (thanks for the excellent work!). Its purpose is to integrate `kw_bd`, the implementation of a real database management system that uses SQLite underneath, into the project. In other words, wherever metadata files are being used to simulate a `DB` or wherever `DB`'s functionalities are advantageous, this integration should happen. Also, this PR adds a script to migrate old metadata files into the `DB`.

**[TODOs]** At the moment this PR is just a draft as there are several missing parts of it:
- ~~Update `kernel-config-manager` test suite (integration is done)~~
- ~~FInish integration with `build` and `deploy` (statistics)~~
- ~~Integrate with `pomodoro`~~
- ~~Integrate with `report`~~
- ~~Fix migration script (migrate the same legacy user data only once)~~

**[TESTING THIS PR]**
If you wish to test this PR (thanks a lot!), below are the steps and the assertions you should make. The TLDR is that just by pulling this PR, installing the PR version of kw and asserting the migration is consistent using `kw report` and `kw kernel-config-manager` is of great help! The steps below are a detailed script to fully test the PR (besides automated tests).

1. It is **REALLY IMPORTANT** to make the backup of your kw personal data. Either use `kw backup <path_to_backup>` or `cp -r <kw_data_dir_path> <path_to_backup>`
2. Have the `unstable` branch updated with `git pull origin unstable`
3. Checkout to the `unstable` branch with `git checkout unstable`
4. Pull this PR locally with `git pull origin pull/836/head:<name_of_branch>`
5. Checkout to the local PR branch with `git switch <name_of_branch>`
6. Install this version of kw with `./setup.sh --install`. You should see a warning like `Creating database: <kw_data_dir_path>/kw.db` and no errors
7. Assert that the file `<kw_data_dir_path>/kw.db` has been created
8. Assert that in your `<kw_data_dir_path>`, the directories `pomodoro` and `statistics` are renamed to `legacy_pomodoro` and `legacy_statistics`, respectively. For the kernel configs, assert that in your `<kw_data_dir_path>/configs`, the directories `configs` and `metadata` are renamed to `legacy_configs` and `legacy_metadata` and that the actual kernel config files (the kernel `.config`s) are in `<kw_data_dir_path>/configs`
9. Run `kw report <period of time>` of a period of time that you can assert the output
10. Experiment with `kw build`, `kw deploy`, `kw deploy --list`, `kw deploy --uninstall` or `kw deploy --modules-deploy` and assert through `kw report --statistics` their addition
11. Experiment with `kw pomodoro --set-timer <timebox>` and assert through `kw report --pomodoro` their addition. Also use `kw pomodoro --check-timer` to see if the active timeboxes are displayed correctly
12. Experiment with `kw kernel-config-manager` and assert that the feature works as intended
13. Reinstall this version of kw with `./setup.sh --install`. You should **NOT** see a warning like `Creating database: <kw_data_dir_path>/kw.db` (also, no errors should happen)
14. Assert that `<kw_data_dir_path>` hasn't changed and that there is no duplicated data. You could use `sqlite3 <kw_data_dir_path>/kw.db` and check that the tables are consistent (e.g. `SELECT * FROM statistics_report;`) or juts use `kw report`
15. If any corruption of your data has happened, restore your backup (recommend deleting the contents of `<kw_data_dir_path>` and running `kw backup --restore <path_to_backup> --force`)
16. THANK YOU SO MUCH FOR TESTING THIS PR!

### CODE REVIEW

**Comment by davidbtadokoro** [comment]
**[UPDATE]**
This force push adds a commit that greatly refactors the `kernel-config-manager` test suite (in the `tests/kernel_config_manager_test.sh` file). Besides making the code more clean and updated with the project code style, the objective was to make it easier to alter it for the `kw_db` integration (which is the next step in this PR).

**Comment by davidbtadokoro** [comment]
**[UPDATE]**
These last couple force pushes made more codestyle and refactoring updates to the `kw kernel-config-manager` test suite (in commit `tests: kernel_config_manager_test: Update codestyle and refactor structure`) and updated the test suite for the `kw_db` integration.

**Comment by davidbtadokoro** [comment]
**[UPDATE]**
This force push adds the integration with `build` and `deploy`. But there are some fixes needed to be done before consider it finished. The commit [src: kwlib: Integrate with kw_db](https://github.com/kworkflow/kworkflow/pull/836/commits/de8d14d91b8d0ba5d2b8a3ceb6883f0bfdbb8404) will be amended with the rest of the changes.

**Comment by davidbtadokoro** [comment]
**[UPDATE]**
This force push finishes the `build` and `deploy` (statistics) integration.

**Comment by davidbtadokoro** [comment]
**[UPDATE]**
Last force-push integrates `pomodoro`.

File: src/kw_db.sh [hunk 1]
```diff
@@ -152,8 +152,8 @@ function format_values_db()
 
   for val in "$@"; do
     [[ "$count" -eq 0 ]] && values+='('
-    # check if not a sqlite function
-    if [[ ! "$val" =~ ^[[:alnum:]_]+\(.*\)$ ]]; then
+    # check if not a sqlite function nor NULL
```
Comment (rodrigosiqueira) [inline src/kw_db.sh:246]: Could you expand this function documentation to add one example about this 'correct format'?

File: database/kwdb.sql [hunk 2]
```diff
@@ -0,0 +1,222 @@
+PRAGMA foreign_keys = 1;
+BEGIN TRANSACTION;
+-- Tables
+
+-- This table holds the metadata of the config files handled by kw
+CREATE TABLE IF NOT EXISTS "config" (
```
Comment (rodrigosiqueira) [inline database/kwdb.sql:6]: Rename it to `kernel_config`.

File: database/kwdb.sql [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  "id" INTEGER NOT NULL UNIQUE,
+  "name" TEXT NOT NULL UNIQUE,
+  "description" TEXT,
+  "path" TEXT NOT NULL UNIQUE,
```
Comment (rodrigosiqueira) [inline database/kwdb.sql:10]: Maybe we add a field for tracking when the config file was saved.

File: database/kwdb.sql [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  "name" TEXT NOT NULL UNIQUE,
+  PRIMARY KEY("id")
+);
+
+-- Table containing user created tags
+CREATE TABLE IF NOT EXISTS "tag" (
+  "id" INTEGER NOT NULL UNIQUE,
+  "name" TEXT NOT NULL UNIQUE,
+  "active" INTEGER NOT NULL CHECK ("active" IN (0, 1)) DEFAULT 1,
+  PRIMARY KEY("id")
```
Comment (rodrigosiqueira) [inline database/kwdb.sql:34]: Could you add a comment that describes this field in more detail?

File: database/kwdb.sql [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+INSERT OR IGNORE INTO "status" ("name") VALUES
+  ('success'),
+  ('failure'),
+  ('interrupted'),
+  ('unknown');
+
+INSERT OR IGNORE INTO "command_label" ("name") VALUES
+  ('build'),
+  ('codestyle'),
+  ('configm'),
```
Comment (rodrigosiqueira) [inline database/kwdb.sql:85]: Update this command, it should be `kernel_config_manager`.

File: database/kwdb.sql [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+  ('deploy'),
+  ('device'),
+  ('drm'),
+  ('explore'),
+  ('list'),
+  ('maintainers'),
+  ('modules_deploy'),
+  ('ssh'),
+  ('uninstall');
```
Comment (rodrigosiqueira) [inline database/kwdb.sql:94]: Please, update this list with all the command options (e.g., VM, self-update, etc).

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by davidbtadokoro** [comment]
**[UPDATE]**
Last force-push integrates with `report`. Still need to make the changes requested above.

File: database/migrate_legacy_data_20220101.sh [hunk 7]
```diff
@@ -0,0 +1,226 @@
+#!/bin/bash
+
+# This file handles the migration of user data, from the old folders and files
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:3]: /user data/of legacy user data/

File: database/migrate_legacy_data_20220101.sh [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+# storage, to the new database model
+
+KW_LIB_DIR='src'
+KW_DB_DIR='database'
+. 'src/kw_include.sh' --source-only
+include "$KW_LIB_DIR/kwio.sh"
+include "$KW_LIB_DIR/kwlib.sh"
+include "$KW_LIB_DIR/kw_string.sh"
+include "$KW_LIB_DIR/kw_db.sh"
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh]: Please, review the code style in this entire file to ensure it follows kw standard.

File: database/migrate_legacy_data_20220101.sh [hunk 9]
```diff
@@ -0,0 +1,226 @@
+#!/bin/bash
+
+# This file handles the migration of user data, from the old folders and files
+# storage, to the new database model
+
+KW_LIB_DIR='src'
+KW_DB_DIR='database'
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:7]: Let's make it readonly and move it after the include.

File: database/migrate_legacy_data_20220101.sh [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+  # users have lots of entries
+  while [[ "$#" -gt 0 ]]; do
+    rows="$(format_values_db "$length" "${@:1:$n}")"
+    insert_into "$table" "$columns" "$rows"
+
+    shift "$n"
+  done
+}
+
+function db_migration()
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:192]: How about `db_migration_main()`? Move it to the beginning of this file.

File: database/migrate_legacy_data_20220101.sh [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+{
+  execute_sql_script "$KW_DB_DIR/kwdb.sql"
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:194]: Do you know if this command always passes? If not, we need to check its return.

File: database/migrate_legacy_data_20220101.sh [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+  local start_date
+  local label_name
+  local status
+  local elapsed_time_in_secs
+  local rows
+  local columns='("date","label_name","status","elapsed_time_in_secs")'
+  local count=0
+  local -a values=()
+
+  file_list=$(find "${datadir}/statistics" -type f | sort)
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:32]: Use `--dictionary-order` after `sort`.

File: database/migrate_legacy_data_20220101.sh [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+  for file in $file_list; do
+    # This line converts the file path into a usable date value
+    # E.g.: ${datadir}/statistics/2021/10/05 -> 2021-10-05
+    start_date=$(printf '%s\n' "$file" | sed -e 's/.*statistics\///' -e 's/\//-/g')
+    while IFS=' ' read -r label_name elapsed_time_in_secs; do
+      status='success'
+      # The new database stores execution status, making this label unnecessary
+      if [[ "$label_name" == 'build_failure' ]]; then
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:41]: We also need to handle `deploy_failure`.

File: database/migrate_legacy_data_20220101.sh [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        rows="$(format_values_db 4 "${values[@]}")"
+        insert_into '"statistics_report"' "$columns" "$rows"
+
+        values=()
+        count=0
+      fi
+    done < "$file"
+  done
+
+  # insert_values '"statistics_report"' "$columns" 4 "${values[@]}"
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:60]: We can drop it.

File: database/migrate_legacy_data_20220101.sh [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+        rows="$(format_values_db 5 "${values[@]}")"
+        insert_into '"pomodoro_report"' "$columns" "$rows"
+
+        values=()
+        count=0
+      fi
+    done < "$file"
+  done
+
+  # insert_values '"pomodoro_report"' "$columns" 5 "${values[@]}"
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:114]: Drop this line.

File: database/migrate_legacy_data_20220101.sh [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+  local -a file_list
+  local line
+  local name
+  local path
+  local description
+  local rows
+  local columns='("name","description","path")'
+  local count=0
+  local -a values=()
+
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:135]: drop this extra line.

File: database/migrate_legacy_data_20220101.sh [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+  local configs_dir="${datadir}/configs/configs"
+  local metadata_dir="${datadir}/configs/metadata"
+
+  file_list=$(find "$configs_dir" -type f | sort)
+
+  for file in $file_list; do
+    path="$file"
+    name="${path##*/}" # get just the file name
+    description=$(cat "${metadata_dir}/${name}")
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:144]: Check: https://kworkflow.org/content/codingstyle.html#command-substitution-and-arithmetic-expression

File: database/migrate_legacy_data_20220101.sh [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+    if [[ "$count" -ge 100 ]]; then
+      rows="$(format_values_db 3 "${values[@]}")"
+      insert_into '"config"' "$columns" "$rows"
+
+      values=()
+      count=0
+    fi
+  done
+
+  # insert_values '"config"' "$columns" 3 "${values[@]}"
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:161]: Drop this line.

File: database/migrate_legacy_data_20220101.sh [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+      ;;
+    s)
+      : # Do nothing
+      ;;
+  esac
+
+  printf '%s\n' "$time_value"
+}
+
+function insert_values()
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:171]: Can we delete this function?

File: setup.sh [hunk 20]
```diff
@@ -365,12 +367,20 @@ function synchronize_files()
   cmd_output_manager "rsync -vr $CONFIG_DIR/ $etcdir" "$verbose"
   ASSERT_IF_NOT_EQ_ZERO "The command 'rsync -vr $CONFIG_DIR/ $etcdir $verbose' failed" "$?"
 
+  # Database files
+  mkdir -p "$databasedir"
+  cmd_output_manager "rsync -vr $DATABASE/ $databasedir" "$verbose"
```
Comment (rodrigosiqueira) [inline setup.sh:372]: Remember to use the complete parameter name.

File: setup.sh [hunk 21]
```diff
@@ -365,12 +367,20 @@ function synchronize_files()
...
+
   setup_global_config_file
 
   # User data
   mkdir -p "$datadir"
   mkdir -p "$datadir/statistics"
   mkdir -p "$datadir/configs"
+  if [[ -x "$databasedir/migrate_legacy_data_20220101.sh" ]]; then
+    eval "$databasedir/migrate_legacy_data_20220101.sh"
+  fi
```
Comment (rodrigosiqueira) [inline setup.sh:383]: How will we set up the empty database if we don't have the migration script?

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

File: tests/kw_db_test.sh [hunk 22]
```diff
@@ -261,4 +261,121 @@ function test_select_from()
...
+  output=$(replace_into table columns '')
+  ret="$?"
+  expected='Empty table or rows.'
+  assert_equals_helper 'Empty rows, error expected' "$LINENO" 22 "$ret"
+  assert_equals_helper 'Wrong error message' "$LINENO" "$expected" "$output"
+
+  # valid operation with non-existent row
+  replace_into 'fake_table' '("name","attribute1","attribute2")' "('someName','someAtt1','someAtt2')"
+  ret="$?"
+  output=$(sqlite3 "$KW_DATA_DIR/kw.db" -batch "SELECT 'f'.'name' FROM 'fake_table' AS 'f' WHERE 'f'.'name'='someName' ;")
```
Comment (rodrigosiqueira) [inline tests/kw_db_test.sh:292]: https://kworkflow.org/content/codingstyle.html#string-concatenation

check the other parts of this test.

File: src/kwlib.sh [hunk 23]
```diff
@@ -1,5 +1,6 @@
 # NOTE: src/kw_config_loader.sh must be included before this file
 include "$KW_LIB_DIR/kw_string.sh"
+include "$KW_LIB_DIR/kw_db.sh"
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:3]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: src/kwlib.sh [hunk 24]
```diff
@@ -338,62 +339,35 @@ function detect_distro()
...
-  day_path="$KW_DATA_DIR/statistics/$year_month_dir/$day"
-
-  elapsed_time=$(date -d@"$value" -u +%H:%M:%S)
+  local label_name="$1"
+  local elapsed_time_in_secs="$2"
+  local status=${3:-'success'}
+  local start_date
+  local start_time
+  local row
+  local columns='("label_name","status","date","time","elapsed_time_in_secs")'
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:356]: How about `database_columns`?

File: src/kwlib.sh [hunk 25]
```diff
@@ -338,62 +339,35 @@ function detect_distro()
...
+
+  start_date=$(date +%Y-%m-%d)
+  start_time=$(date +%H:%M:%S)
+
+  elapsed_time=$(date -d@"$elapsed_time_in_secs" -u +%H:%M:%S)
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:361]: Replace `-u` with `--utc` and `-d` with `--date=`.

File: tests/kwlib_test.sh [hunk 26]
```diff
@@ -6,20 +6,19 @@ include './tests/utils.sh'
...
 function oneTimeSetUp()
 {
-  KW_DATA_DIR="$SHUNIT_TMPDIR"
-  TARGET_YEAR_MONTH="2020/05"
-  FAKE_STATISTICS_PATH="$KW_DATA_DIR/statistics"
-  FAKE_STATISTICS_MONTH_PATH="$FAKE_STATISTICS_PATH/$TARGET_YEAR_MONTH"
-  FAKE_STATISTICS_DAY_PATH="$FAKE_STATISTICS_MONTH_PATH/03"
-  ORIGINAL_DIR="$PWD"
+  declare -gr ORIGINAL_DIR="$PWD"
+  declare -gr FAKE_DATA="$SHUNIT_TMPDIR/db_testing"
```
Comment (rodrigosiqueira) [inline tests/kwlib_test.sh:10]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: database/migrate_legacy_data_20220101.sh [hunk 27]
```diff
@@ -8,6 +8,7 @@ KW_DB_DIR='database'
 . 'src/kw_include.sh' --source-only
 include "$KW_LIB_DIR/kwio.sh"
 include "$KW_LIB_DIR/kwlib.sh"
+include "$KW_LIB_DIR/kw_time_and_date.sh"
```
Comment (rodrigosiqueira) [inline database/migrate_legacy_data_20220101.sh:11]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: src/kw_time_and_date.sh [hunk 28]
```diff
@@ -1,3 +1,5 @@
+include "$KW_LIB_DIR/kw_string.sh"
```
Comment (rodrigosiqueira) [inline src/kw_time_and_date.sh:1]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: src/kw_db.sh [hunk 29]
```diff
@@ -209,7 +212,11 @@ function select_from()
     return 22 # EINVAL
   fi
 
-  sqlite3 -init "$KW_DB_DIR/pre_cmd.sql" -cmd "$pre_cmd" "$db_path" -batch "SELECT $columns FROM $table;"
+  query="SELECT $columns FROM $table ;"
+  if [[ -n "${order_by}" ]]; then
+    query="SELECT $columns FROM $table ORDER BY ${order_by} ;"
+  fi
+  sqlite3 -init "$KW_DB_DIR/pre_cmd.sql" -cmd "$pre_cmd" "$db_path" -batch "$query"
```
Comment (rodrigosiqueira) [inline src/kw_db.sh:219]: Can we just use `select_from`?

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

File: database/kwdb.sql [hunk 30]
```diff
@@ -0,0 +1,222 @@
...
+  "e_id" AS "id",
+  "name" AS "label_name",
+  "status_name" AS "status",
+  "date",
+  "time",
+  "elapsed_time_in_secs"
+FROM
+  (SELECT "id" AS "e_id", "date", "time" FROM "event")
+JOIN "executed" ON "executed"."id" IS "e_id"
+JOIN "command_label" ON "command_label"."id" IS "executed"."command_label_id"
```
[the remark this answers is not in the document]
Comment (davidbtadokoro) [reply database/kwdb.sql:143]: Sorry if I'm understanding it wrong. I think `JOIN` and `INNER JOIN` are equivalent. `NATURAL JOIN` requires the attributes in the `ON` clause to have the same name. The benefit is that, by doing this, the line would be less verbose and more informative for someone that knows what a `NATURAL JOIN` is (e.g. line 142 would be `NATURAL JOIN "executed"`. The downside is that it requires renaming of attributes in the file and doesn't explicitly indicates which attribute (or if there are more than one) is used in the join condition. Imo, I think keeping it this way is better. Any thoughts?

File: tests/kernel_config_manager_test.sh [hunk 31]
```diff
@@ -96,26 +117,24 @@ function test_save_config_file_check_directories_creation()
...
 
 function test_save_config_file_check_saved_config()
 {
   local output
 
   save_config_file "$NO_FORCE" "$NAME_1" "$DESCRIPTION_1" > /dev/null
-
-  assertTrue "${LINENO}: Failed to find ${NAME_1}" '[[ -f ${dot_configs_dir}/configs/${NAME_1} ]]'
-  assertTrue "${LINENO}: Failed the metadata related to ${NAME_1}" '[[ -f $dot_configs_dir/metadata/${NAME_1} ]]'
+  assertTrue "${LINENO}: Failed to find .config file for ${NAME_1}" '[[ -f ${dot_configs_dir}/${NAME_1} ]]'
```
[the remark this answers is not in the document]
Comment (davidbtadokoro) [reply tests/kernel_config_manager_test.sh:128]: Sorry, I didn't catch it.

File: src/kw_db.sh [hunk 32]
```diff
@@ -95,10 +95,93 @@ function insert_into()
   sqlite3 -init "$KW_DB_DIR/pre_cmd.sql" "$db_path" -batch "INSERT INTO $table $entries VALUES $values;"
 }
 
+# This function updates or insert rows into table of given database,
```
[the remark this answers is not in the document]
Comment (davidbtadokoro) [reply src/kw_db.sh:98]: Got it!

File: src/kw_db.sh (hunk 29 above)
Comment (davidbtadokoro) [reply src/kw_db.sh:219]: Sorry if I'm  missing something, but I don't think we can do this here as this is part of the definition of `select_from` itself.

File: src/kw_db.sh (hunk 32 above)
[the remark this answers is not in the document]
Comment (davidbtadokoro) [reply src/kw_db.sh:98]: Done it!

File: database/migrate_legacy_data_20220101.sh [hunk 33]
```diff
@@ -0,0 +1,232 @@
...
+  local label_name
+  local status
+  local elapsed_time_in_secs
+  local rows
+  local columns='("date","label_name","status","elapsed_time_in_secs")'
+  local count=0
+  local -a values=()
+
+  # if there are no statistics reports to migrate, return
+  [[ ! -d "${datadir}/statistics" ]] && return 0
```
Comment (davidbtadokoro) [inline database/migrate_legacy_data_20220101.sh:59]: Mark migrated directories to not duplicate data.

**Comment by rodrigosiqueira** [comment]
I merged this change in the unstable branch.

Thanks a lot @kwy95 @davidbtadokoro @magalilemes for all the hard work. This is a fantastic change that put kw at another level :)

**Pull request closed without merging**